rotator/drain.go

```diff
@@ -0,0 +1,579 @@
+	// any finished pod can be removed
+	if pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed {
+		return true, nil, nil
+	}
+
+	controllerRef := getPodController(pod)
+	if controllerRef != nil {
+		return true, nil, nil
+	}
+	if o.Force {
+		return true, &warning{kUnmanagedWarning}, nil
+	}
+
+	return false, nil, &fatal{kUnmanagedFatal}
+}
+
+type DaemonSetFilterOptions struct {
+	client           typedappsv1.AppsV1Interface
+	force            bool
+	ignoreDaemonSets bool
+}
+
+func (o *DaemonSetFilterOptions) daemonSetFilter(pod corev1.Pod) (bool, *warning, *fatal) {
+	ctx := context.TODO()
+
+	// Note that we return false in cases where the pod is DaemonSet managed,
+	// regardless of flags.  We never delete them, the only question is whether
+	// their presence constitutes an error.
+	//
+	// The exception is for pods that are orphaned (the referencing
+	// management resource - including DaemonSet - is not found).
+	// Such pods will be deleted if Force is used.
+	controllerRef := getPodController(pod)
+	if controllerRef == nil || controllerRef.Kind != "DaemonSet" {
+		return true, nil, nil
+	}
+
+	if _, err := o.client.DaemonSets(pod.Namespace).Get(ctx, controllerRef.Name, metav1.GetOptions{}); err != nil {
+		// remove orphaned pods with a warning if Force is used
+		if apierrors.IsNotFound(err) && o.force {
+			return true, &warning{err.Error()}, nil
+		}
+		return false, nil, &fatal{err.Error()}
+	}
+
+	if !o.ignoreDaemonSets {
+		return false, nil, &fatal{kDaemonsetFatal}
+	}
+
+	return false, &warning{kDaemonsetWarning}, nil
+}
+
+func mirrorPodFilter(pod corev1.Pod) (bool, *warning, *fatal) {
+	if _, found := pod.ObjectMeta.Annotations[corev1.MirrorPodAnnotationKey]; found {
+		return false, nil, nil
+	}
+	return true, nil, nil
+}
+
+func hasLocalStorage(pod corev1.Pod) bool {
+	for _, volume := range pod.Spec.Volumes {
+		if volume.EmptyDir != nil {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (o *DrainOptions) localStorageFilter(pod corev1.Pod) (bool, *warning, *fatal) {
+	if !hasLocalStorage(pod) {
+		return true, nil, nil
+	}
+	if !o.DeleteLocalData {
+		return false, nil, &fatal{kLocalStorageFatal}
+	}
+	return true, &warning{kLocalStorageWarning}, nil
+}
+
+// Map of status message to a list of pod names having that status.
+type podStatuses map[string][]string
+
+func (ps podStatuses) message() string {
+	msgs := []string{}
+
+	for key, pods := range ps {
+		msgs = append(msgs, fmt.Sprintf("%s: %s", key, strings.Join(pods, ", ")))
+	}
+	return strings.Join(msgs, "; ")
+}
+
+// getPodsForDeletion receives resource info for a node, and returns all the pods from the given node that we
+// are planning on deleting. If there are any pods preventing us from deleting, we return that list in an error.
+func getPodsForDeletion(client kubernetes.Interface, node *corev1.Node, options *DrainOptions) ([]corev1.Pod, error) {
```

Comment:
@gabrieljackson drain was the most complex part of the job because its not supported natively by k8s. This code here is a combination of open shift and k8s code adapted to our needs 😅 . Missed the returns somehow, should be good now.